### ISSUE TITLE
Consider reporting skipped tests in the json reporter

### ISSUE THREAD

**Issue by natebosch** [dart-lang/test#1321]
We currently report skipped tests as success for backwards compatibility. I would imagine than consumers of this data can do something smarter if we report skipped tests correctly.

Rolling out a change here might be a little tricky since the breaking change in semver doesn't inform things like IDE plugin compatibility.

cc @DanTup for thoughts. Would this be useful?

https://github.com/dart-lang/test/blob/223665cec34bb90ce09a6475907efaa8e2c886e1/pkgs/test_core/lib/src/runner/reporter/json.dart#L253-L256

**Comment by natebosch** [comment]
Oh, we _do_ report them as skipped, just in a separate field right below this...

maybe it's not worth changing if the only real benefit is that we can send 1 fewer field.

**Comment by DanTup** [comment]
Yeah, in Dart-Code we have this:

```ts
		if (evt.skipped) {
			testNode.status = TestStatus.Skipped;
		} else if (evt.result === "success") {
			testNode.status = TestStatus.Passed;
		}
```

If you're looking for things to improve though, those fake "Loading" tests that come through like real tests are kinda annoying :-)

**Comment by natebosch** [comment]
> If you're looking for things to improve though, those fake "Loading" tests that come through like real tests are kinda annoying :-)

We have a `hidden` field on `testDone` events. I don't know why we have one there but not on `testStart` events.

From the docs:

https://github.com/dart-lang/test/blob/fbbd3b2381fd9a713e3149e30a2cc1a88a2a5e9d/pkgs/test/doc/json_reporter.md#L302-L305

The implementation doesn't make a lot of sense to me:

https://github.com/dart-lang/test/blob/fbbd3b2381fd9a713e3149e30a2cc1a88a2a5e9d/pkgs/test_core/lib/src/runner/reporter/json.dart#L258

Are you currently reading that `hidden` field? Is it useful for you?

Would it be better to suppress "Loading" tests always, or would it be better to be more precise about what `hidden` means and add that field to `testStart` events?

**Comment by jakemac53** [comment]
In general we do want to report the loading status of tests, etc. That is useful information for tools to get.

The way that works is pretty bizarre currently to be sure though.

**Comment by DanTup** [comment]
> Are you currently reading that `hidden` field? Is it useful for you?

I don't think so. I might not have seen it (since it was only on the end one). Currently we check `testNode.test.name.startsWith("loading ")` to try and identify them and hide them.

> Would it be better to suppress "Loading" tests always, or would it be better to be more precise about what hidden means and add that field to testStart events?

I think either of these could simplify things a fair bit, though having some indication of loading would definitely be better (so if we can come up with a nice way to show it in the test tree, we could).


> The way that works is pretty bizarre currently to be sure though.

Yeah. I assumed that pre-JSON it was probably convenient to emit the tests that way (to show loading status) and then it just didn't translate very nicely to the JSON. I think having some specific event for "progress" events might be better though.
